packages/protocol/contracts/identity/Random.sol

```diff
@@ -93,20 +95,31 @@ contract Random is IRandom, Ownable, Initializable {
    */
   function addRandomness(uint256 blockNumber, bytes32 randomness) internal {
     history[blockNumber] = randomness;
-    if (historySize == 0) {
-      historyFirst = block.number;
-      historySize = 1;
-    } else if (historySize > randomnessBlockRetentionWindow) {
-      delete history[historyFirst];
-      delete history[historyFirst + 1];
-      historyFirst += 2;
-      historySize--;
-    } else if (historySize == randomnessBlockRetentionWindow) {
-      delete history[historyFirst];
-      historyFirst++;
+    if ((blockNumber + 1) % getEpochSize() == 0) {
+      if (lastEpochBlock != historyFirst) {
```

Comment:
Interesting, haven't thought about it. If the retention window is actually larger than the epoch size, then we are in trouble too? I.e. we should only delete `if lastEpochBlock < historyFirst`?

Reply:
True. The code around deleting seems very fishy to me. Good catch